File: CTCI/ch1/oneAway.py
'''
	Given two strings, write a function to check if they are one edit (or zero edits) away. An edit
	is considered inserting a character, removing a character, or replacing a character.
'''

def lengthDifference(string1, string2):
	return abs(len(string1) - len(string2))

def checkReplaceOrRemove(string1, string2):
	#string1 is the bigger string
	x = 0
	y = 0
	editsMade = 0
	while x < len(string1) and y < len(string2):
		if string1[x] != string2[y]:
			editsMade+=1
			x+=1
		else:
			x+=1
			y+=1
	return editsMade


def oneAway(inputString1, inputString2):
	diff = lengthDifference(inputString1, inputString2)
	editsMade = 0
	if diff > 1:
		return False	
	# both strings are the samee size
	elif diff == 0:
		for x in range(0, len(inputString1)):
			if inputString1[x] != inputString2[x]:
				editsMade+=1
	# strings differ in length by 1
	elif diff == 1:
		if len(inputString1) > len(inputString2):
			editsMade = checkReplaceOrRemove(inputString1, inputString2)
		else:
			editsMade = checkReplaceOrRemove(inputString2, inputString1)
	if editsMade > 1:
		return False
	return True



print(oneAway('pale', 'ple'))
print(oneAway('pales', 'pale'))
print(oneAway('pale', 'bale'))
print(oneAway('pale', 'bake'))
print(oneAway('pizza', 'italy'))
print(oneAway('bus', 'backpack'))

'''
	prints:
		> True
		> True
		> True
		> False
		> False
		> False
'''

File: CTCI/ch1/oneAway.swift
func lengthDifference(string1 : String, string2 : String) -> Int {
    let s1Length : Int = string1.characters.count;
    let s2Length : Int = string2.characters.count;
    let diff : Int = s1Length - s2Length;
    return diff < 0 ? diff * -1 : diff;
}

func checkEdits(s1 : String, s2 : String) -> Int {
    var x : Int = 0;
    var y : Int = 0;
    var editsMade : Int = 0;
    let s1Chars : [Character] = [Character](s1.characters);
    let s2Chars : [Character] = [Character](s2.characters);
    while(x < s1.characters.count && y < s2.characters.count){
        if(s1Chars[x] != s2Chars[y]){
            editsMade+=1;
            x+=1;
        } else {
            x+=1;
            y+=1;
        }
    }
    return editsMade;
}

func oneAway(string1 : String, string2 : String) -> Bool {
    let diff : Int = lengthDifference(string1: string1, string2: string2);
    let s1Chars : [Character] = [Character](string1.characters);
    let s2Chars : [Character] = [Character](string2.characters);
    var editsMade : Int = 0;
    if(diff > 1) {
        return false;
    }
    else if(diff == 0){
        for i : Int in 0 ..< string1.characters.count {
            if(s1Chars[i] != s2Chars[i]){
                editsMade+=1;
            }
        }
    }
    else if(diff == 1){
        if(s1Chars.count > s2Chars.count){
            editsMade = checkEdits(s1: string1, s2: string2);
        } else {
           editsMade = checkEdits(s1: string2, s2: string1)
        }
    }
    
    if editsMade > 1{
        return false;
    }
    return true;
}


print(oneAway(string1: "pale", string2: "ple"));
print(oneAway(string1: "pales", string2: "pale"));
print(oneAway(string1: "pale", string2: "bale"));
print(oneAway(string1: "pale", string2: "bake"));
print(oneAway(string1: "pizza", string2: "italy"));
print(oneAway(string1: "bus", string2: "backpack"));


/*
    > swift oneAway.swift

    prints:
        > True
        > True
        > True
        > False
        > False
        > False
*/